### PR TITLE
[WC-2175]: Gallery Multi Selection

### PR DESCRIPTION
**Pull request by samuelreichert**
### Pull request type

New feature (non-breaking change which adds functionality)

---

### Description

Adding a new feature to Gallery: You can use the keyboard to select multiple items inside the gallery, just as simple as shift + arrow keys. You can also unselect an item with shift+space.
* Shift + arrow up
* Shift + arrow down
* Shift + arrow left
* Shift + arrow right
* Shift + Home
* Shift + End
* Shift + Page Up
* Shift + Page Down

### CODE REVIEW

File: packages/pluggableWidgets/gallery-web/src/components/ListItem.tsx [hunk 1]
```diff
@@ -17,8 +17,7 @@ type ListItemProps = Omit<JSX.IntrinsicElements["div"], "ref" | "role"> & {
     eventsController: ItemEventsController;
 };
 
-// eslint-disable-next-line prefer-arrow-callback
-const component = memo(function ListItem(props: ListItemProps): ReactElement {
+export function ListItem(props: ListItemProps): ReactElement {
```
Comment (iobuhov) [inline packages/pluggableWidgets/gallery-web/src/components/ListItem.tsx:20]: Why `memo` is removed?

Comment (samuelreichert) [reply packages/pluggableWidgets/gallery-web/src/components/ListItem.tsx:20]: @iobuhov The list item selection (classes) were not kept up to date with the actual keyboard selection

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts [hunk 2]
```diff
@@ -277,3 +296,22 @@ function objectListEqual(a: ObjectItem[], b: ObjectItem[]): boolean {
     const setB = new Set(b.map(obj => obj.id));
     return a.every(obj => setB.has(obj.id));
 }
+
+export type Positions = {
+    columnIndex: number;
+    rowIndex: number;
+};
+
+function getRowIndex(index: number, numberOfColumns: number): number {
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:305]: 4. (NIT) Rename this funciton to `getRowBasedOnItemIndex` and add jsdoc comment explainging what is doing.

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts [hunk 3]
```diff
@@ -277,3 +296,22 @@ function objectListEqual(a: ObjectItem[], b: ObjectItem[]): boolean {
...
+    return Math.floor(index / numberOfColumns);
+}
+
+export function getPosition(numberOfColumns: number, totalItems: number, index: number): Positions {
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:309]: 5. (NIT) Rename this funciton to `getColumnAndRowBasedOnIndex` and add jsdoc comment explainging what is doing.

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts [hunk 4]
```diff
@@ -277,3 +296,22 @@ function objectListEqual(a: ObjectItem[], b: ObjectItem[]): boolean {
     const setB = new Set(b.map(obj => obj.id));
     return a.every(obj => setB.has(obj.id));
 }
+
+export type Positions = {
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:300]: 6. (NIT) Rename this type to `PositionInGrid` (or something with grid in name).

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:1]: 7. As we're about to introduce a new method, it's a good idea to do the same thing for `SelectActionHelper` - go to the implementation and add a new method `onSelectAdjacentGrid`.

**Review by iobuhov — CHANGES_REQUESTED** [review]
All in all, PR looks good, but I think we need to split the code into methods to avoid misuse and overload - methods should be simple.

File: packages/pluggableWidgets/gallery-web/src/components/ListItem.tsx (hunk 1 above)
Comment (iobuhov) [reply packages/pluggableWidgets/gallery-web/src/components/ListItem.tsx:20]: Ah, I see, but then can we split this component into "container" and "view" and keep the view part wrapped with in memo?

File: packages/shared/widget-plugin-grid/src/selection/types.ts
[the remark this answers is not in the document]
Comment (samuelreichert) [reply packages/shared/widget-plugin-grid/src/selection/types.ts:1]: In this case I will need both `size` and `numberOfColumns`. For arrow navigation, I still need size, but for the other commands ( "page up", "home", ...) then I can use numberOfColumns.

**Review by leonardomendix — DISMISSED** [review]

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts [hunk 5]
```diff
@@ -277,3 +306,51 @@ function objectListEqual(a: ObjectItem[], b: ObjectItem[]): boolean {
...
+    rowIndex: number;
+};
+
+/**
+ * Given an index and number of columns,
+ * this function returns which row the index is located in a grid.
+ * @function getRowBasedOnItemIndex
+ * @param {number} index - The index used to find the row.
+ * @param {number} numberOfColumns - The number of columns of the grid.
+ * @return {number}
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:321]: Wow, nice 👍

File: packages/shared/widget-plugin-grid/src/selection/helpers.ts [hunk 6]
```diff
@@ -113,15 +113,14 @@ export class MultiSelectionHelper {
...
+            this.selectionValue.setSelection(itemsToAdd);
             return;
         }
 
         const itemsToRemove = this._getRange(start, currentEnd);
         const itemsToAdd = this._getRange(start, nextEnd);
 
-        let selection = [...this.selectionValue.selection];
+        let selection: ObjectItem[] = [];
         selection = this._diff(selection, itemsToRemove);
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/helpers.ts:124]: There one little thing that I'm afraid of - what about checkboxes in datagrid?
I suspect this change will brake current behavior. But, let's see. I'll approve.
@leonardomendix, can you please test Data Grid as well and report any unusual stfuff?

Comment (leonardomendix) [reply packages/shared/widget-plugin-grid/src/selection/helpers.ts:124]: Already checked and it doesn't seem to be affected. Thanks for the heads up.

**Review by gjulivan — DISMISSED** [review]

**Review by leonardomendix — DISMISSED** [review]

**Review by leonardomendix — APPROVED** [review]

**Review by gjulivan — DISMISSED** [review]

**Review by leonardomendix — APPROVED** [review]

File: packages/shared/widget-plugin-grid/src/selection/__tests__/helpers.spec.ts [hunk 7]
```diff
@@ -37,7 +37,7 @@ describe("MultiSelectionHelper", () => {
             expect(gridState()).toEqual([x, _, x, _, x, _, _, _]);
 
             helper.selectUpTo(items[7]);
-            expect(gridState()).toEqual([x, _, x, _, x, x, x, x]);
+            expect(gridState()).toEqual([_, _, _, _, x, x, x, x]);
```
Comment (iobuhov) [inline packages/shared/widget-plugin-grid/src/selection/__tests__/helpers.spec.ts:40]: Can we move this test out of this block?
We have new behavior that don't match previous. Can we add just one more block to explain new behavior?

**Review by iobuhov — CHANGES_REQUESTED** [review]

**Review by iobuhov — APPROVED** [review]
LGTM

**Pull request merged**